pkg/kubernetes/watcher.go

```diff
@@ -0,0 +1,213 @@
+)
+
+type ObjListWatcher struct {
+	// Lock to syncronize the collector update with the watcher
+	Mx *sync.Mutex
+
+	k8sCli       *kubernetes.Clientset
+	ResourceKind string
+	informer     cache.SharedInformer
+	stopChannel  chan struct{}
+
+	// ContainersMetrics holds all container energy and resource usage metrics
+	ContainersMetrics *map[string]*collector_metric.ContainerMetrics
+}
+
+func newK8sClient() *kubernetes.Clientset {
+	var restConf *rest.Config
+	var err error
+	if config.KubeConfig == "" {
+		// creates the in-cluster config
+		restConf, err = rest.InClusterConfig()
+		klog.Infoln("Using in cluster k8s config")
+	} else {
+		// use the current context in kubeconfig
+		restConf, err = clientcmd.BuildConfigFromFlags("", config.KubeConfig)
+		klog.Infoln("Using out cluster k8s config: ", config.KubeConfig)
+	}
+	if err != nil {
+		klog.Infoln("%v", err)
+		return nil
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(restConf)
+	if err != nil {
+		klog.Fatalf("%v", err)
+	}
+	return clientset
+}
+
+func NewObjListWatcher() *ObjListWatcher {
+	w := &ObjListWatcher{
+		stopChannel:  make(chan struct{}),
+		k8sCli:       newK8sClient(),
+		ResourceKind: podResourceType,
+	}
+	if w.k8sCli == nil {
+		return w
+	}
+
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fmt.Sprintf("spec.nodeName=%s", collector_metric.NodeName) // to filter events per node
+	}
+	objListWatcher := cache.NewFilteredListWatchFromClient(
+		w.k8sCli.CoreV1().RESTClient(),
+		w.ResourceKind,
+		metav1.NamespaceAll,
+		optionsModifier,
+	)
+
+	w.informer = cache.NewSharedInformer(objListWatcher, nil, 0)
+	w.stopChannel = make(chan struct{})
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			w.handleUpdate(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.handleDeleted(obj)
+		},
+	})
+	IsWatcherEnabled = true
+	return w
+}
+
+func (w *ObjListWatcher) Run() {
+	if !IsWatcherEnabled {
+		return
+	}
+	go w.informer.Run(w.stopChannel)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		klog.Fatalf("watcher timed out waiting for caches to sync")
+	}
+}
+
+func (w *ObjListWatcher) Stop() {
+	close(w.stopChannel)
+}
+
+func (w *ObjListWatcher) handleUpdate(obj interface{}) {
+	switch w.ResourceKind {
+	case podResourceType:
+		pod, ok := obj.(*k8sv1.Pod)
+		if !ok {
+			klog.Infof("Could not convert obj: %v", w.ResourceKind)
+			return
+		}
+		podID := string(pod.GetUID())
+		// Pod object can have many updates such as change in the annotations and labels.
+		// We only add the pod information when all containers are ready, then when the
+		// pod is in our managed list we can skip the informantion update.
+		if _, exist := managedPods[podID]; exist {
```

Comment:
There is an issue here for the new-coming pod. 

For the early read,  even if you have line `condition.Type != k8sv1.ContainersReady && condition.Status != k8sv1.ConditionTrue ` to check but it turns out that container ID has not been assigned yet. 
I have to remove this check to have my new pod added by the watcher.

```
# I added a following log line to see an update 
# klog.Infof("Watcher add new containerID: %s (pod: %s)", containerID, pod.Name)

I0511 08:53:31.124701       1 watcher.go:181] Watcher add new containerID:  (pod: coremark-cpeh-3068050960-k5qgr)
I0511 08:53:33.844029       1 watcher.go:181] Watcher add new containerID: 9aab3a72841724834469a04dd9c770f7b001bb3c38d6ad046ee13097e17db8ed (pod: coremark-cpeh-3068050960-k5qgr)
```

You can notice that the first assign of managedPods containerID is empty. You may try adding empty Check in addition to the above condition.

Reply:
Thanks @sunya-ch for finding this issue! 
I updated the PR, could you please test it again?

Reply:
Still there is missing system process detection. I'm still under debugging.

```bash
# I added the following debugging log
#		if isSystemProcess {
#			klog.V(5).Infof("System process command: %s (cGroup ID %v)", comm, ct.CGroupID)
#		}
kubectl logs kepler-exporter-vdjlc -n kepler|grep "System process command"|grep coremark
I0511 09:27:00.438820       1 container_hc_collector.go:143] System process command: coremark-16thre (cGroup ID 0)
I0511 09:27:03.455934       1 container_hc_collector.go:143] System process command: coremark-16thre (cGroup ID 0)
```

Reply:
@sunya-ch it might be related to the function `extractPodContainerIDfromPath` in the file pkg/cgroup/resolve_container.go:L297

Reply:
My current problem is that some processes is still missing (not even be processed in the `updateBPFMetrics` function). For the above issue, I think you updated changes look good to me.